hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2017,3 +2032,22 @@ func (r *NodePoolReconciler) doesCPODecompressAndDecodeConfig(ctx context.Contex
 	_, managesDecompressAndDecodeConfig := supportutil.ImageLabels(controlPlaneOperatorImageMetadata)[controlPlaneOperatorManagesDecompressAndDecodeConfig]
 	return managesDecompressAndDecodeConfig, nil
 }
+
+func (r *NodePoolReconciler) ensureMachineDeletion(nodePool *hyperv1.NodePool, controlPlaneNamespace string) error {
+	machines := capiv1.MachineList{}
+	if err := r.List(context.Background(), &machines, &client.ListOptions{Namespace: controlPlaneNamespace}); err != nil {
+		return fmt.Errorf("failed to list MachineSet: %w", err)
+	}
+
+	// Filter out only machines belonging to deleted NodePool
+	var machineSetOwnedMachines []capiv1.Machine
+	for i, machine := range machines.Items {
+		if machine.Annotations[nodePoolAnnotation] == "clusters/"+nodePool.GetName() {
```

Comment:
this name is not guaranteed "clusters" is your particular ns

Reply:
You could also check for existence of
machineDeployment(nodePool *hyperv1.NodePool, controlPlaneNamespace string)
or
machineSet(nodePool *hyperv1.NodePool, controlPlaneNamespace string)
And then filter machines by ownerRef.

Reply:
Understood about clusters not being a guarantee. I will looking into changing this tomorrow.

I sent a Slack message in a thread, but when I was testing, I couldn't look at the ownerRef because the MachineSet name returned [here](https://github.com/bryan-cox/hypershift/blob/910c2419e355f2564eff0fc52cb87c78b3862fff/hypershift-operator/controllers/nodepool/nodepool_controller.go#L873) is not the actual MachineSet name so then I can't compare the ownerRef. It didn't seem like the MachineSet name is exactly identical to the NodePool name. 

For example, when I was testing, my NodePool name was _brcox-hypershift-x86-test-us-east-1a_ but the MachineSet name was _brcox-hypershift-x86-test-us-east-1a-bff454f9f_. I was going to also test tomorrow to see if this [function](https://github.com/bryan-cox/hypershift/blob/910c2419e355f2564eff0fc52cb87c78b3862fff/hypershift-operator/controllers/nodepool/nodepool_controller.go#L810) is even working based on this.

Reply:
I think I found a way that works now.

Reply:
>I couldn't look at the ownerRef because the MachineSet name returned [here](https://github.com/bryan-cox/hypershift/blob/910c2419e355f2564eff0fc52cb87c78b3862fff/hypershift-operator/controllers/nodepool/nodepool_controller.go#L873)

That's because that MachineSet is the one created explicitly for one you set in-place upgrade typer.

You need to account for both scenarios: replace (MachineDeployment known name - which generates machineSets underneath)
and inplace (MachineSet known name)